File: lib/features/news/domain/entities/news_item.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'news_item.freezed.dart';
part 'news_item.g.dart';

@freezed
// ignore_for_file: invalid_annotation_target
class NewsItem with _$NewsItem {
  const factory NewsItem({
    @JsonKey(readValue: _readId) required String id,
    required String title,
    required String content,
    @Default('none') String mediaType,
    String? mediaPath,
    String? thumbnailPath,
    @Default(false) bool isPublished,
    @Default('published') String status,
    DateTime? publishedAt,
    required DateTime createdAt,
    required DateTime updatedAt,
    @JsonKey(readValue: _readAuthorId) required String authorId,
    @JsonKey(readValue: _readAuthor) Map<String, dynamic>? author,
    // ─── Metadata ───
    @Default('Announcements') String category,
    @Default(false) bool isPinned,
    // ─── Reaction System ───
    @Default([]) List<Map<String, dynamic>> reactions,
    @Default({}) Map<String, dynamic> reactionCounts,
    // ─── View Tracking ───
    @Default([]) List<String> viewers,
    @Default(0) int viewCount,
  }) = _NewsItem;

  factory NewsItem.fromJson(Map<String, dynamic> json) => _$NewsItemFromJson(json);
}

/// Reads 'id' first, then falls back to '_id' for MongoDB compatibility.
Object? _readId(Map<dynamic, dynamic> json, String key) {
  return json['id'] ?? json['_id'];
}

/// When Mongoose populates authorId, it becomes an object { _id, fullName }.
/// This extracts the _id string in that case.
Object? _readAuthorId(Map<dynamic, dynamic> json, String key) {
  final val = json['authorId'];
  if (val is Map) return val['_id']?.toString() ?? '';
  return val?.toString() ?? '';
}

/// When Mongoose populates authorId, extract it as the author map.
/// Falls back to a dedicated 'author' field if present.
Object? _readAuthor(Map<dynamic, dynamic> json, String key) {
  final authorId = json['authorId'];
  if (authorId is Map) {
    return Map<String, dynamic>.from(authorId);
  }
  return json['author'];
}

File: server/models/NewsItem.js
const mongoose = require('mongoose');

const NewsItemSchema = new mongoose.Schema({
    title: {
        type: String,
        required: [true, 'Заголовок обязателен'],
        trim: true,
        maxlength: [200, 'Заголовок не может быть длиннее 200 символов']
    },
    content: {
        type: String,
        required: [true, 'Содержание обязательно'],
        trim: true
    },
    mediaType: {
        type: String,
        enum: ['none', 'image', 'video'],
        default: 'none'
    },
    mediaPath: {
        type: String,
        default: null
    },
    thumbnailPath: {
        type: String,
        default: null
    },
    isPublished: {
        type: Boolean,
        default: false
    },
    authorId: {
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User',
        required: true
    },
    // ─── News Metadata ───
    category: {
        type: String,
        enum: ['Academic', 'Announcements', 'Events', 'Urgent'],
        default: 'Announcements',
        required: true
    },
    isPinned: {
        type: Boolean,
        default: false
    },
    // ─── Reaction System ───
    reactions: [{
        userId: { type: mongoose.Schema.Types.ObjectId, ref: 'User' },
        type: { type: String, enum: ['heart', 'thumbsUp', 'fire', 'party'] }
    }],
    reactionCounts: {
        heart: { type: Number, default: 0 },
        thumbsUp: { type: Number, default: 0 },
        fire: { type: Number, default: 0 },
        party: { type: Number, default: 0 }
    },
    // ─── View Tracking ───
    viewers: [{
        type: mongoose.Schema.Types.ObjectId,
        ref: 'User'
    }],
    viewCount: {
        type: Number,
        default: 0
    }
}, {
    timestamps: true
});

// Index for feed queries (published news, newest first)
// Pinned first, then sorted by createdAt
NewsItemSchema.index({ isPublished: 1, isPinned: -1, createdAt: -1 });
NewsItemSchema.index({ title: 'text' });

module.exports = mongoose.model('NewsItem', NewsItemSchema);
